### PR TITLE
Initial version of the SQL adapter

### LINKED ISSUE #1: Create initial version of the SQL adapter
#### Description:
The initial prototype of the SQL adapter should have the following capabilities:
	
- Pushing and Pulling objects to/from an existing table. Only primitive properties (i.e. number, string, bool,...) are supported. This means complex objects types that require a collection of linked tables are not supported. CustomObjects, on the other hand, are supported as long as the content of their CustomData matches the column of the SQL table.
- Ability to associate a table with a BHoM type (information is stored in a special table of the database itself). This means :
  - Rows pulled from a "type-strong' table will automatically be converted to objects of that type
  - If a type is associated with a single table, the user can push objects of that type without having to specify the target table.

### PR DESCRIPTION
**Pull request by adecler**
### Issues addressed by this PR

Closes #1

See issue for more details

The table that links the tables to types has to be named `_tableTypes` and should contains two properties as follows:

image

This is only required if you want your tables to behave in a strong-typed manner but will work without it or if only some of your table are type-strong. 

### Test files
https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/SQL_Toolkit/%231-InitialVersion?csf=1&web=1&e=9az7l5

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@BHoMBot check project compliance

**Comment by adecler** [comment]
@BHoMBot check project-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

File: SQL_oM/SQL_oM.csproj
Comment (FraserGreenroyd) [inline SQL_oM/SQL_oM.csproj:38]:
```suggestion
    <Reference Include="BHoM">
      <SpecificVersion>False</SpecificVersion>
      <HintPath>C:\ProgramData\BHoM\Assemblies\BHoM.dll</HintPath>
    </Reference>
```

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix copyright headers ref. 2087696624

**Comment by FraserGreenroyd** [comment]
@BHoMBot fix project file ref. 2087696090

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by al-fisher** [comment]
Thanks for updating the SQL_oM.csproj error @FraserGreenroyd 🙏   - was just looking at that

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

**Review by al-fisher — CHANGES_REQUESTED** [review]
This is looking good. Tested pushing and pulling to various dummy tables - and behaving as expected.
One minor issue @adecler - the Push returns `success false` even when it was indeed successful.
Will continue to test some more

**Comment by adecler** [comment]
@BHoMBot check compliance

**Comment by adecler** [comment]
@BHoMBot check core

**Comment by adecler** [comment]
@BHoMBot check core

**Comment by adecler** [comment]
@BHoMBot check compliance

**Comment by tg359** [comment]
Me and @michaelhoehn had a go at pushing some random stuff to some tables and encountered the following issues. I'm not sure whether these should be full Issue issues, or could be handled easier here:

- Deserialisation of BHoM_Guid from the SQL DB fails as the field name BHoM_Guid  is being stored as a string in the DB, and the custom object being created from the Pull is being given the `string` version not a `Guid`. Could this be made to automatically convert the BHoM_Guid field into a Guid dtype?
- Can a type-strong table be created from a set of single dtype objects? Something like an option in the pushConfig to create the table if it doesn't exist would be useful for setup (along with the associate methods to make sure the attributes/fields match if it already exists).

**Comment by adecler** [comment]
@tg359, @michaelhoehn ,

Thanks for checking out the SQL adapter 😄 .

- I have checked that Guid can be pushed and pulled correctly and they are indeed working as expected BUT you have to make sure their datatype is defined as `uniqueidentifier` in the schema, not as a string.

- It is definitely possible to automatically create tables if they don't already exist. If you look at the code from this PR, you will notice a few `CreateTable` methods that are doing just that. The key question is "should we ?". We already have MongoDB as a database that can accept any type of data. And the Mongo adapter will also create the collection if it doesn't already exist. The SQL database is meant for a type different type of use cases: storing specific results (with curated schema) with the purpose the visualise/explore/communicate those results. With that in mind, we expect most of the objects pushed to SQL being of the `IResult` type and specifically designed for being stored in a database. That means, for example, flat data (i.e. only primitive properties) that can easily be processed by something like PowerBI. So, I am not saying we will never facilitate the creation of tables. But I would like it to be a manual, conscious, and controlled process for now (until we have use cases that justify more flexibility).

I hope that all make sense. I have created a document to collect our user stories [here](https://burohappold.sharepoint.com/:f:/r/sites/Direction/designtechnology/RandDWishlist/00217_SQL%20Adapter?csf=1&web=1&e=WTW02y). Feel free to add your use cases to it.

**Comment by adecler** [comment]
@FraserGreenroyd , @al-fisher , could you update your review of this PR, please ?

**Review by FraserGreenroyd — APPROVED** [review]
This looks fine from my testing now, sorry for the delay @adecler . I believe @tg359 has been testing more as well in addition to your testing :smile:

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core

**Comment by adecler** [comment]
Thanks @FraserGreenroyd :thumb:

@al-fisher , your review is still blocking the PR despite your request being addressed (see commit on the 12th of March). Could you update your review please ?

@tg359 , it would be great if you could also provide a review before this is merged.

**Comment by al-fisher** [comment]
Perfect 
Thanks @adecler 
Yes - have dismissed my review.

**Review by tg359 — APPROVED** [review]
Tested using the sample script and this worked for me.

**Pull request merged**
